ext/js/dictionary/dictionary-database.js

```diff
@@ -347,6 +392,98 @@ export class DictionaryDatabase {
+        }
+        // otherwise, you are the worker, so do the work
+        safePerformance.mark('drawMedia:start');
+
+        // merge items with the same path to reduce the number of database queries. collects the canvases into a single array for each path.
+        /** @type {Map<string, import('dictionary-database').DrawMediaGroupedRequest>} */
+        const groupedItems = new Map();
+        for (const item of items) {
+            const {path, dictionary, canvasIndex, canvasWidth, canvasHeight, generation} = item;
+            const key = `${path}:::${dictionary}`;
+            if (!groupedItems.has(key)) {
+                groupedItems.set(key, {path, dictionary, canvasIndexes: [], canvasWidth, canvasHeight, generation});
+            }
+            groupedItems.get(key)?.canvasIndexes.push(canvasIndex);
+        }
+        const groupedItemsArray = [...groupedItems.values()];
+
+        /** @type {import('dictionary-database').FindPredicate<import('dictionary-database').MediaRequest, import('dictionary-database').MediaDataArrayBufferContent>} */
+        const predicate = (row, item) => (row.dictionary === item.dictionary);
+        const results = await this._findMultiBulk('media', ['path'], groupedItemsArray, this._createOnlyQuery5, predicate, this._createDrawMediaBind);
+
+        // move all svgs to front to have a hotter loop
+        results.sort((a, _b) => (a.mediaType === 'image/svg+xml' ? -1 : 1));
+
+        safePerformance.mark('drawMedia:draw:start');
+        for (const m of results) {
+            if (m.mediaType === 'image/svg+xml') {
+                safePerformance.mark('drawMedia:draw:svg:start');
+                /** @type {import('@resvg/resvg-wasm').ResvgRenderOptions} */
+                const opts = {
+                    fitTo: {
+                        mode: 'width',
+                        value: m.canvasWidth,
+                    },
+                    font: {
+                        fontBuffers: this._resvgFontBuffer !== null ? [this._resvgFontBuffer] : [],
+                    },
+                };
+                const resvgJS = new Resvg(new Uint8Array(m.content), opts);
+                const render = resvgJS.render();
+                source.postMessage({action: 'drawBufferToCanvases', params: {buffer: render.pixels.buffer, width: render.width, height: render.height, canvasIndexes: m.canvasIndexes, generation: m.generation}}, [render.pixels.buffer]);
+                safePerformance.mark('drawMedia:draw:svg:end');
+                safePerformance.measure('drawMedia:draw:svg', 'drawMedia:draw:svg:start', 'drawMedia:draw:svg:end');
+            } else {
+                safePerformance.mark('drawMedia:draw:raster:start');
+
+                // ImageDecoder is slightly faster than Blob/createImageBitmap, but
+                // 1) it is not available in Firefox <133
+                // 2) it is available in Firefox >=133, but it's not possible to transfer VideoFrames cross-process
```

Comment:
If we want to use ImageDecoder, yes, as it returns a VideoFrame (the reason its VideoFrame is because of GIF support). And ImageDecoder would be ideal as it can be a simple unified and performant interface for doing this.

But we'll need to wait for Firefox to support cross-process transfer of more things with postMessage, which I don't expect to come any time soon :crying_cat_face: